modules/common/Riot.py

```diff
@@ -51,32 +54,28 @@ def run_riot(self, owl_file, dir_output, json_file, owl_jq):
         """
         path_output = os.path.join(dir_output, json_file)
         # Set JVM memory limits
-        run_env = os.environ.copy()
-        run_env["_JAVA_OPTIONS"] = "-Xms4096m -Xmx8192m"
+        self.configure_jvm()
+        riot_jq_cmd_string = (
+            f"{self.riot_cmd} -v --output JSON-LD {owl_file} | "
+            f"{self.jq_cmd} -r '{owl_jq}' > {path_output}"
+        )
         try:
-            with open(path_output, "wb") as json_output, \
-                    subprocess.Popen([self.riot_cmd, "--output", "JSON-LD", owl_file], env=run_env,
-                                     stdout=subprocess.PIPE,
-                                     stderr=subprocess.PIPE) as riot_process, \
-                    subprocess.Popen([self.jq_cmd, "-r", owl_jq], env=run_env,
-                                     stdin=riot_process.stdout,
-                                     stdout=subprocess.PIPE,
-                                     stderr=subprocess.PIPE) as jq_process:
-                json_output.write(jq_process.stdout.read())
-                logger.error(f"RIOT stderr -> '{riot_process.stderr.read()}'")
-                logger.error(f"JQ stderr -> '{jq_process.stderr.read()}'")
-        except EnvironmentError as e:
-            msg_err = "When running RIOT for OWL file '{}', "
-            "with destination path '{}' and JQ filter '{}', "
-            "the following error occurred: '{}'".format(owl_file, path_output, owl_jq, e)
-            logger.error(msg_err)
-            raise RiotException(msg_err)
-        except Exception as e:
-            msg_err = "When running RIOT for OWL file '{}', "
-            "with destination path '{}' and JQ filter '{}', "
-            "the following error occurred: '{}'".format(owl_file, path_output, owl_jq, e)
+            riot_jq = subprocess.run(
+                riot_jq_cmd_string,
+                env=os.environ.copy(),
+                shell=True,
+                text=True,
+                capture_output=True,
+                check=False,
+            )
+            riot_jq.check_returncode()
+        except subprocess.CalledProcessError as err:
+            msg_err = f"The following error occurred: '{err}'"
             logger.error(msg_err)
             raise RiotException(msg_err)
+        if "error" in riot_jq.stderr.lower():
```

Comment:
This one is a little bit fishy 😏 ...
So, the problem we've faced when running this RIOT+JQ tandem has a different nature depending on the tool.

When running RIOT for some of the ontologies, the default heap space in the JVM was not enough for processing some _.owl_ files, so it started crashing for those cases, this was fixed by a difficult try+error on how to actually get RIOT to accept the JVM parameters for an bigger heap space, but, we didn't actually look at whether it was returning any error code back to the shell, mainly due to the way RIOT was wired into PIS and connected to JQ.

8GiB of heap space turned out to be more than enough to keep processing _.owl_ files, so nowadays we don't really have memory problems with RIOT, but we need a reliable method to detect whether it fails or not, so we can communicate that through the _manifest file_.

Now, in which cases would it fail?
- When it faces an _.owl_ file conversion that requires more than 8GiB. This is very unlikely, as it was ok with the default heap space size, which I think is about 1 or 2 GiB
- When running PIS in a VM on GCP with less than 8GiB RAM
- When the _.owl_ file it's processing is wrong, e.g. bad formatting, it's not an _.owl_ file, it's empty...

Would you say the proposed code covers these _error scenarios_?

Reply:
Would you mind testing it for the last 2 scenarios on a VM in GCP?

Reply:
The second part of the problem -> JQ

_JQ_ is another tool that we use in two different scenarios, both of them destined to filter/manipulate data in _JSON_ format:
1. Together with RIOT, over converted ontology files
2. Standalone, over whatever _JSON_ file we have collected

As _JQ_ seems to be quite memory hungry, in both use cases we've found ourselves in a situation where the tool's memory usage exceeds the memory in the machine, so the OS kills the process but, PIS doesn't catch that, despite all the exception handling that was wired into those _subprocess.Popen_ calls

Unfortunately, when the OS kills the running _JQ_ subprocess, no _stderr_ content is produced, it is a silent event, which means checking _stderr_ would not catch a vanishing _JQ_ subprocess.

To find out how to catch when _JQ_ fails due to insufficient memory, in a way that you focus just on _JQ_, you can use one of the use cases where _JQ_ is used on its own, i.e. not with RIOT, on a GCP VM with very little memory to quickly trigger the error situation.

I would recommend doing it on a separated short Python script, to not having to deal with all the boilerplate around running PIS.